Shader.cpp: GCC6 compile error

```
In file included from /usr/include/x86_64-linux-gnu/c++/6/bits/c++allocator.h:33:0,
                 from /usr/include/c++/6/bits/allocator.h:46,
                 from /usr/include/c++/6/string:41,
                 from /home/jochen/git-source/inviwo/modules/opengl/include/modules/opengl/inviwoopengl.h:51,
                 from /home/jochen/git-source/inviwo/modules/opengl/include/modules/opengl/shader/shader.h:34,
                 from /home/jochen/git-source/inviwo/modules/opengl/src/shader/shader.cpp:30:
/usr/include/c++/6/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>; _Args = {const std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>&}; _Tp = std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>]’:
/usr/include/c++/6/bits/alloc_traits.h:475:4:   required from ‘static void std::allocator_traits<std::allocator<_CharT> >::construct(std::allocator_traits<std::allocator<_CharT> >::allocator_type&, _Up*, _Args&& ...) [with _Up = std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>; _Args = {const std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>&}; _Tp = std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>; std::allocator_traits<std::allocator<_CharT> >::allocator_type = std::allocator<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment> >]’
/usr/include/c++/6/bits/hashtable_policy.h:1953:37:   required from ‘std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type* std::__detail::_Hashtable_alloc<_NodeAlloc>::_M_allocate_node(_Args&& ...) [with _Args = {const std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>&}; _NodeAlloc = std::allocator<std::__detail::_Hash_node<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>, true> >; std::__detail::_Hashtable_alloc<_NodeAlloc>::__node_type = std::__detail::_Hash_node<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>, true>]’
/usr/include/c++/6/bits/hashtable_policy.h:180:58:   required from ‘std::__detail::_AllocNode<_NodeAlloc>::__node_type* std::__detail::_AllocNode<_NodeAlloc>::operator()(_Arg&&) const [with _Arg = const std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>&; _NodeAlloc = std::allocator<std::__detail::_Hash_node<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>, true> >; std::__detail::_AllocNode<_NodeAlloc>::__node_type = std::__detail::_Hash_node<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>, true>]’
/usr/include/c++/6/bits/hashtable.h:1681:18:   required from ‘std::pair<typename std::__detail::_Hashtable_base<_Key, _Value, _ExtractKey, _Equal, _H1, _H2, _Hash, _Traits>::iterator, bool> std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::_M_insert(_Arg&&, const _NodeGenerator&, std::true_type) [with _Arg = const std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>&; _NodeGenerator = std::__detail::_AllocNode<std::allocator<std::__detail::_Hash_node<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>, true> > >; _Key = inviwo::ShaderType; _Value = std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>; _Alloc = std::allocator<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment> >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<inviwo::ShaderType>; _H1 = std::hash<inviwo::ShaderType>; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<true, false, true>; typename std::__detail::_Hashtable_base<_Key, _Value, _ExtractKey, _Equal, _H1, _H2, _Hash, _Traits>::iterator = std::__detail::_Node_iterator<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>, false, true>; std::true_type = std::integral_constant<bool, true>]’
/usr/include/c++/6/bits/hashtable_policy.h:713:55:   required from ‘std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__ireturn_type std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::insert(const value_type&) [with _Key = inviwo::ShaderType; _Value = std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>; _Alloc = std::allocator<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment> >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<inviwo::ShaderType>; _H1 = std::hash<inviwo::ShaderType>; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<true, false, true>; std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__ireturn_type = std::pair<std::__detail::_Node_iterator<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>, false, true>, bool>; std::__detail::_Insert_base<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::value_type = std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>]’
/usr/include/c++/6/bits/unordered_map.h:550:31:   required from ‘std::pair<typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__detail::__is_noexcept_hash<_Key, _Hash> > >::value, false, true> >::iterator, bool> std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::insert(const value_type&) [with _Key = inviwo::ShaderType; _Tp = inviwo::Shader::ShaderAttachment; _Hash = std::hash<inviwo::ShaderType>; _Pred = std::equal_to<inviwo::ShaderType>; _Alloc = std::allocator<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment> >; typename std::_Hashtable<_Key, std::pair<const _Key, _Tp>, _Alloc, std::__detail::_Select1st, _Pred, _Hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<std::__not_<std::__and_<std::__is_fast_hash<_Hash>, std::__detail::__is_noexcept_hash<_Key, _Hash> > >::value, false, true> >::iterator = std::__detail::_Node_iterator<std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>, false, true>; std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::value_type = std::pair<const inviwo::ShaderType, inviwo::Shader::ShaderAttachment>]’
/home/jochen/git-source/inviwo/modules/opengl/src/shader/shader.cpp:111:95:   required from here
/usr/include/c++/6/ext/new_allocator.h:120:4: error: use of deleted function ‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = const inviwo::ShaderType; _T2 = inviwo::Shader::ShaderAttachment]’
  { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
                 from /usr/include/c++/6/bits/char_traits.h:39,
                 from /usr/include/c++/6/string:40,
                 from /home/jochen/git-source/inviwo/modules/opengl/include/modules/opengl/inviwoopengl.h:51,
                 from /home/jochen/git-source/inviwo/modules/opengl/include/modules/opengl/shader/shader.h:34,
                 from /home/jochen/git-source/inviwo/modules/opengl/src/shader/shader.cpp:30:
/usr/include/c++/6/bits/stl_pair.h:288:17: note: ‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = const inviwo::ShaderType; _T2 = inviwo::Shader::ShaderAttachment]’ is implicitly deleted because the default definition would be ill-formed:
       constexpr pair(const pair&) = default;
                 ^~~~
/usr/include/c++/6/bits/stl_pair.h:288:17: error: use of deleted function ‘inviwo::Shader::ShaderAttachment::ShaderAttachment(const inviwo::Shader::ShaderAttachment&)’
In file included from /home/jochen/git-source/inviwo/modules/opengl/src/shader/shader.cpp:30:0:
/home/jochen/git-source/inviwo/modules/opengl/include/modules/opengl/shader/shader.h:62:9: note: declared here
         ShaderAttachment(const ShaderAttachment &) = delete;
         ^~~~~~~~~~~~~~~~
```

--- a/README.md
+++ b/README.md
@@ -19,7 +19,7 @@ Below follow some general information about the framework:
  - The core is written in C/C++ only, with minor dependencies.
  
 #### Modules
- - Modern graphics programming through OpenGL (>= 3.2 Core) related modules.
+ - Modern graphics programming through OpenGL (>= 3.3 Core) related modules.
  - Parallel computing on multiple platforms (GPU/CPU) through OpenCL (>= 1.0) related modules.
  - Python (> 3.2) scripting and computation is supported through provided modules.
 

--- a/include/inviwo/core/util/capabilities.h
+++ b/include/inviwo/core/util/capabilities.h
@@ -41,16 +41,9 @@ public:
     virtual ~Capabilities();
 
     virtual void printInfo() = 0;
-
-    virtual bool canAllocate(glm::u64 dataSize, glm::u8 percentageOfAvailableMemory = 100);
-
-    virtual uvec3 calculateOptimalBrickSize(uvec3 dimensions, size_t formatSizeInBytes,
-                                            glm::u8 percentageOfAvailableMemory = 100);
-
     virtual void retrieveStaticInfo() = 0;
 
 protected:
-    virtual glm::u64 getMemorySizeInBytes(uvec3 dimensions, size_t formatSizeInBytes);
     virtual void retrieveDynamicInfo() = 0;
 };
 

--- a/include/inviwo/core/util/systemcapabilities.h
+++ b/include/inviwo/core/util/systemcapabilities.h
@@ -81,10 +81,6 @@ public:
 
     int numberOfCores() const;
 
-    virtual bool canAllocate(glm::u64 dataSize, glm::u8 percentageOfAvailableMemory = 100) override;
-    virtual uvec3 calculateOptimalBrickSize(uvec3 dimensions, size_t formatSizeInBytes,
-                                            glm::u8 percentageOfAvailableMemory = 100) override;
-
     glm::u64 getAvailableMemory();
     glm::u64 getCurrentResidentMemoryUsage();
 

--- a/modules/opengl/include/modules/opengl/openglcapabilities.h
+++ b/modules/opengl/include/modules/opengl/openglcapabilities.h
@@ -44,12 +44,12 @@ public:
         GLSLShaderVersion(int num);
         GLSLShaderVersion(int num, std::string pro);
 
-        std::string getProfile();
-        int getVersion();
-        std::string getVersionAsString();
-        std::string getVersionAndProfileAsString();
+        const std::string& getProfile() const;
+        int getVersion() const;
+        std::string getVersionAsString() const;
+        std::string getVersionAndProfileAsString() const;
 
-        bool hasProfile();
+        bool hasProfile() const;
 
         static bool sortHighestFirst(GLSLShaderVersion i, GLSLShaderVersion j);
 
@@ -65,10 +65,7 @@ public:
 
     virtual void printInfo() override;
     void printDetailedInfo();
-    
-    virtual bool canAllocate(glm::u64 dataSize, glm::u8 percentageOfAvailableMemory = 100) override;
-    virtual uvec3 calculateOptimalBrickSize(uvec3 dimensions, size_t formatSizeInBytes,
-                                            glm::u8 percentageOfAvailableMemory = 100) override;
+
     static void initializeGLEW();
     static int getOpenGLVersion();
 
@@ -80,35 +77,35 @@ public:
     static bool isExtensionSupported(const char*);
     static bool isSupported(const char*);
 
-    bool isTexturesSupported();
-    bool is3DTexturesSupported();
-    bool isTextureArraysSupported();
-    bool isFboSupported();
-    bool isShadersSupported();
-    bool isShadersSupportedARB();
-    bool isGeometryShadersSupported();
+    bool isTexturesSupported() const;
+    bool is3DTexturesSupported() const;
+    bool isTextureArraysSupported() const;
+    bool isFboSupported() const;
+    bool isShadersSupported() const;
+    bool isShadersSupportedARB() const;
+    bool isGeometryShadersSupported() const;
 
     GLSLShaderVersion getCurrentShaderVersion();
-    std::string getCurrentGlobalGLSLHeader();
-    std::string getCurrentGlobalGLSLVertexDefines();
-    std::string getCurrentGlobalGLSLFragmentDefines();
+    size_t getCurrentShaderIndex() const;
+    size_t getNumberOfShaderVersions() const;
+    GLSLShaderVersion getShaderVersion(size_t ind) const;
 
     glm::u64 getCurrentAvailableTextureMem();
     glm::u64 getTotalAvailableTextureMem();
 
-    std::string getRenderString() const;
-    std::string getVendorString() const;
-    std::string getGLVersionString() const;
-    std::string getGLSLVersionString() const;
+    const std::string& getRenderString() const;
+    const std::string& getVendorString() const;
+    const std::string& getGLVersionString() const;
+    const std::string& getGLSLVersionString() const;
     GlVendor getVendor() const;
 
-    int getMaxProgramLoopCount();
-    int getMaxTexSize();
-    int getMax3DTexSize();
-    int getMaxArrayTexSize();
-    int getMaxArrayVertexAttribs();
-    int getMaxColorAttachments();
-    int getNumTexUnits();
+    int getMaxProgramLoopCount() const;
+    int getMaxTexSize() const;
+    int getMax3DTexSize() const;
+    int getMaxArrayTexSize() const;
+    int getMaxArrayVertexAttribs() const;
+    int getMaxColorAttachments() const;
+    int getNumTexUnits() const;
 
     static std::string getPreferredProfile();
     bool setPreferredProfile(std::string, bool);
@@ -116,10 +113,6 @@ public:
 protected:
     virtual void retrieveStaticInfo() override;
     virtual void retrieveDynamicInfo() override;
-
-    void rebuildGLSLHeader();
-    void rebuildGLSLVertexDefines();
-    void rebuildGLSLFragmentDefines();
 
     void addShaderVersion(GLSLShaderVersion);
     void addShaderVersionIfEqualOrLower(GLSLShaderVersion, int);
@@ -142,9 +135,9 @@ private:
     void print(const Ts&... messages) const {
         std::stringstream ss;
         printHelper(ss, messages...);
-        LogCentral::getPtr()->log("OpenGLInfo", LogLevel::Info, LogAudience::User, "", "", 0, ss.str());
+        LogCentral::getPtr()->log("OpenGLInfo", LogLevel::Info, LogAudience::User, "", "", 0,
+                                  ss.str());
     }
-
 
     static bool glewInitialized_;
     static std::string preferredProfile_;
@@ -167,9 +160,6 @@ private:
     int geometryShadersMaxTotalOutputComponents_;
 
     size_t currentGlobalGLSLVersionIdx_;
-    std::string currentGlobalGLSLHeader_;
-    std::string currentGlobalGLSLVertexDefines_;
-    std::string currentGlobalGLSLFragmentDefines_;
     std::vector<GLSLShaderVersion> supportedShaderVersions_;
 
     // Texturing
@@ -185,6 +175,6 @@ private:
     int numTexUnits_;
 };
 
-}  // namespace
+}  // namespace inviwo
 
 #endif  // IVW_OPENGLCAPABILITIES_H

--- a/modules/opengl/include/modules/opengl/shader/shader.h
+++ b/modules/opengl/include/modules/opengl/shader/shader.h
@@ -153,6 +153,8 @@ public:
     void removeOnReload(const BaseCallBack *callback);
 
 private:
+    void bindAttributes();
+
     void verify() const;
     void handleError(OpenGLException &e);
     std::string processLog(std::string log) const;

--- a/modules/opengl/include/modules/opengl/shader/shadermanager.h
+++ b/modules/opengl/include/modules/opengl/shader/shadermanager.h
@@ -62,13 +62,7 @@ public:
     void unregisterShader(Shader* shader);
     bool isRegistered(Shader* shader) const;
 
-
-    std::string getGlobalGLSLHeader();
-    std::string getGlobalGLSLVertexDefines();
-    std::string getGlobalGLSLFragmentDefines();
     int getGlobalGLSLVersion();
-
-    void bindCommonAttributes(unsigned int);
 
     void addShaderSearchPath(std::string);
     const std::vector<std::string>& getShaderSearchPaths();
@@ -89,8 +83,7 @@ public:
     template <typename T>
     std::shared_ptr<Callback> onWillRemoveShader(T&& callback);
 
-protected:
-    OpenGLCapabilities* getOpenGLCapabilitiesObject();
+    OpenGLCapabilities* getOpenGLCapabilities();
 
 private:
     bool addShaderSearchPathImpl(const std::string &);

--- a/modules/opengl/include/modules/opengl/shader/shaderobject.h
+++ b/modules/opengl/include/modules/opengl/shader/shaderobject.h
@@ -39,8 +39,25 @@
 
 namespace inviwo {
 
+class OpenGLCapabilities;
+
 class IVW_MODULE_OPENGL_API ShaderObject {
 public:
+    struct InDeclaration {
+        std::string name;
+        int location = -1;
+        std::string type = "vec4";
+        std::string decl = "in {type} {name};";
+        std::string toString() const;
+    };
+    struct OutDeclaration {
+        std::string name;
+        int location = -1;
+        std::string type = "vec4";
+        std::string decl = "out {type} {name};";
+        std::string toString() const;
+    };
+
     using Callback = std::function<void(ShaderObject*)>;
 
     ShaderObject(ShaderType shaderType, std::shared_ptr<const ShaderResource> resource);
@@ -48,7 +65,7 @@ public:
     ShaderObject(ShaderType shaderType, std::string fileName);
     ShaderObject(std::string fileName);
     ShaderObject(GLenum shaderType, std::string fileName);
-    
+
     ShaderObject(const ShaderObject& rhs);
     ShaderObject(ShaderObject&& rhs) noexcept;
     ShaderObject& operator=(const ShaderObject& that);
@@ -79,41 +96,89 @@ public:
     void clearShaderExtensions();
 
     /**
-     * \brief adds an additional output specifier to the fragment shader
+     * \brief adds an additional output specifier to the shader
      * The given name will be added as
-     * 
-     *     out vec4 __name__;
-     * 
-     * If location index is positive, the output will be
-     * 
-     *     layout(location = __location__) out vec4 __name__;
-     * 
-     * Location indices can be reused several times unless more than
-     * one output specifier is used.
+     *
+     *     out __type__ __name__;
+     *
+     * The shader will call glBindFragDataLocation for each of the output declarations
+     * before linking the shader
      *
      * @param name      identifier of the output specifier
      * @param location  index location of the output (< MAX_RENDER_TARGETS)
+     * @param type      type used for the output specifier
      */
-    void addOutDeclaration(std::string name, int location = -1);
+    void addOutDeclaration(std::string name, int location = -1, const std::string& type = "vec4");
+    void addOutDeclaration(const OutDeclaration& decl);
     void clearOutDeclarations();
+    const std::vector<OutDeclaration>& getOutDeclarations() const;
 
-    std::pair<std::string, unsigned int> resolveLine(size_t line) const;
+    /**
+     * \brief adds an additional input specifier to the shader
+     * The given name will be added as
+     *
+     *     in __type__ __name__;
+     *
+     * The shader will call glBindAttribLocation for each of the input declarations
+     * before linking the shader
+     *
+     * @param name      identifier of the output specifier
+     * @param location  index location of the output (< MAX_RENDER_TARGETS)
+     * @param type      type used for the output specifier
+     */
+    void addInDeclaration(std::string name, int location = -1, const std::string& type = "vec4");
+    void addInDeclaration(const InDeclaration& decl);
+    void clearInDeclarations();
+    const std::vector<InDeclaration>& getInDeclarations() const;
+
+    /**
+     * Adds the default fragment out declarations to the list of out declarations.
+     * This function is automatically called in the constructor for a fragment shader.
+     * If clearOutDeclarations is called then they will be removed and one would have to
+     * manually call this function to re-add them if needed.
+     */
+    void addStandardFragmentOutDeclarations();
+
+    /**
+     * Adds the default vertex in declarations to the list of in declarations.
+     * This function is automatically called in the constructor for a vertex shader.
+     * If clearInDeclarations is called then they will be removed and one would have to
+     * manually call this function to re-add them if needed.
+     */
+    void addStandardVertexInDeclarations();
+
+    std::pair<std::string, size_t> resolveLine(size_t line) const;
     std::string print(bool showSource = false, bool preprocess = true) const;
 
     template <typename T>
     std::shared_ptr<Callback> onChange(T&& callback);
 
 private:
+    struct LineNumberResolver {
+        LineNumberResolver() = default;
+        void addLine(const std::string& file, size_t line);
+        std::pair<std::string, size_t> resolveLine(size_t line) const;
+        void clear();
+        size_t size() { return lines_.size(); }
+        std::string resolveLog(const std::string& compileLog) const;
+        auto begin() const { return lines_.cbegin(); }
+        auto end() const { return lines_.cend(); }
+
+    private:
+        std::vector<std::pair<std::string, size_t>> lines_;
+    };
+
     static std::shared_ptr<const ShaderResource> loadResource(std::string fileName);
     void addDefines(std::ostringstream& source);
-    void addOutDeclarations(std::ostringstream& source);
     void addIncludes(std::ostringstream& source, std::shared_ptr<const ShaderResource> resource);
 
     // state variables
     ShaderType shaderType_;
     GLuint id_;
     std::shared_ptr<const ShaderResource> resource_;
-    std::vector<std::pair<std::string, int>> outDeclarations_;
+
+    std::vector<InDeclaration> inDeclarations_;
+    std::vector<OutDeclaration> outDeclarations_;
 
     using ShaderDefines = std::map<std::string, std::string>;
     ShaderDefines shaderDefines_;
@@ -124,18 +189,17 @@ private:
     // derived variables
     std::string sourceProcessed_;
     std::vector<std::shared_ptr<const ShaderResource>> includeResources_;
-    std::vector<std::pair<std::string, unsigned int> > lineNumberResolver_;
+    LineNumberResolver lnr_;
 
     Dispatcher<void(ShaderObject*)> callbacks_;
     std::vector<std::shared_ptr<ShaderResource::Callback>> resourceCallbacks_;
 };
 
 template <typename T>
-std::shared_ptr<ShaderObject::Callback>
-ShaderObject::onChange(T&& callback) {
-   return callbacks_.add(std::forward<T>(callback));
+std::shared_ptr<ShaderObject::Callback> ShaderObject::onChange(T&& callback) {
+    return callbacks_.add(std::forward<T>(callback));
 }
 
-}  // namespace
+}  // namespace inviwo
 
 #endif  // IVW_SHADEROBJECT_H

--- a/modules/opengl/include/modules/opengl/shader/shaderobject.h
+++ b/modules/opengl/include/modules/opengl/shader/shaderobject.h
@@ -41,6 +41,28 @@ namespace inviwo {
 
 class OpenGLCapabilities;
 
+
+/**
+ * A wrapper for an OpenGL shader object.
+ * Handles loading sources from shader resources, either files or stings.
+ * Pre-processes the sources resolving all include with help of the shader manager.
+ * Keeps a lookup table of from which include each line originates. 
+ * Handles a list of defines, that can be added or removed @see addShaderDefine
+ * Handles a list of shader extensions @see addShaderExtension
+ * Handles a list of input declarations @set addInDeclaration
+ * Handles a list of output declarations @set addOutDeclaration
+ * 
+ * A fragment shader will by default have the following out declarations:
+ *     out vec4 FragData0;   (location 0)
+ *     out vev4 PickingData; (location 1)
+ * and a vertex shader will by default have the following in declarations:
+ *     in vec4 in_Vertex;   (location 0)
+ *     in vec3 in_Normal;   (location 1)
+ *     in vec4 in_Color;    (location 2)
+ *     in vec3 in_TexCoord; (location 3)
+ * The defaults can be removed by calling clearOutDeclarations or clearInDeclarations
+ * respectively.
+ */
 class IVW_MODULE_OPENGL_API ShaderObject {
 public:
     struct InDeclaration {
@@ -135,7 +157,10 @@ public:
      * Adds the default fragment out declarations to the list of out declarations.
      * This function is automatically called in the constructor for a fragment shader.
      * If clearOutDeclarations is called then they will be removed and one would have to
-     * manually call this function to re-add them if needed.
+     * manually call this function to re-add them if needed. 
+     * The defaults are:
+     *     out vec4 FragData0;   (location 0)
+     *     out vev4 PickingData; (location 1)
      */
     void addStandardFragmentOutDeclarations();
 
@@ -144,6 +169,11 @@ public:
      * This function is automatically called in the constructor for a vertex shader.
      * If clearInDeclarations is called then they will be removed and one would have to
      * manually call this function to re-add them if needed.
+     * The defaults are:
+     *     in vec4 in_Vertex;   (location 0)
+     *     in vec3 in_Normal;   (location 1)
+     *     in vec4 in_Color;    (location 2)
+     *     in vec3 in_TexCoord; (location 3)
      */
     void addStandardVertexInDeclarations();
 

--- a/modules/opengl/src/openglcapabilities.cpp
+++ b/modules/opengl/src/openglcapabilities.cpp
@@ -41,6 +41,8 @@
 
 namespace inviwo {
 
+// Or minimal opengl version is 3.3 and glsl version 330
+
 bool OpenGLCapabilities::glewInitialized_ = false;
 std::string OpenGLCapabilities::preferredProfile_ = "core";
 int OpenGLCapabilities::glVersion_ = 0;
@@ -53,19 +55,19 @@ OpenGLCapabilities::GLSLShaderVersion::GLSLShaderVersion(int num) : number_(num)
 OpenGLCapabilities::GLSLShaderVersion::GLSLShaderVersion(int num, std::string pro)
     : number_(num), profile_(pro) {}
 
-std::string OpenGLCapabilities::GLSLShaderVersion::getProfile() { return profile_; }
+const std::string& OpenGLCapabilities::GLSLShaderVersion::getProfile() const { return profile_; }
 
-int OpenGLCapabilities::GLSLShaderVersion::getVersion() { return number_; }
+int OpenGLCapabilities::GLSLShaderVersion::getVersion() const { return number_; }
 
-std::string OpenGLCapabilities::GLSLShaderVersion::getVersionAsString() {
+std::string OpenGLCapabilities::GLSLShaderVersion::getVersionAsString() const {
     return toString<int>(number_);
 }
 
-std::string OpenGLCapabilities::GLSLShaderVersion::getVersionAndProfileAsString() {
+std::string OpenGLCapabilities::GLSLShaderVersion::getVersionAndProfileAsString() const {
     return (hasProfile() ? getVersionAsString() + " " + profile_ : getVersionAsString());
 }
 
-bool OpenGLCapabilities::GLSLShaderVersion::hasProfile() { return (profile_ != ""); }
+bool OpenGLCapabilities::GLSLShaderVersion::hasProfile() const { return (profile_ != ""); }
 
 bool OpenGLCapabilities::GLSLShaderVersion::sortHighestFirst(GLSLShaderVersion i,
                                                              GLSLShaderVersion j) {
@@ -90,14 +92,12 @@ OpenGLCapabilities::OpenGLCapabilities(OpenGLSettings* settings)
     , maxArrayVertexAttribs_(-1)
     , maxColorAttachments_(-1)
     , numTexUnits_(-1) {
+
     supportedShaderVersions_.clear();
-    currentGlobalGLSLHeader_ = "";
-    currentGlobalGLSLVertexDefines_ = "";
-    currentGlobalGLSLFragmentDefines_ = "";
 
     preferredProfile_ = settings->selectedOpenGLProfile_.getSelectedIdentifier();
 
-    settings->btnOpenGLInfo_.onChange([this](){printDetailedInfo();});
+    settings->btnOpenGLInfo_.onChange([this]() { printDetailedInfo(); });
 
     bool hasOutputedGLSLVersionOnce = false;
     settings->selectedOpenGLProfile_.onChange(
@@ -115,8 +115,8 @@ OpenGLCapabilities::OpenGLCapabilities(OpenGLSettings* settings)
         });
 }
 
-OpenGLCapabilities::~OpenGLCapabilities() { 
-    TextureUnit::deinitialize(); 
+OpenGLCapabilities::~OpenGLCapabilities() {
+    TextureUnit::deinitialize();
 
     // reset stuff.
     glewInitialized_ = false;
@@ -200,37 +200,9 @@ void OpenGLCapabilities::printDetailedInfo() {
     }
 }
 
-bool OpenGLCapabilities::canAllocate(glm::u64 dataSize, glm::u8 percentageOfAvailableMemory) {
-    return getCurrentAvailableTextureMem() * percentageOfAvailableMemory / 100 >= dataSize;
-}
-
-uvec3 OpenGLCapabilities::calculateOptimalBrickSize(uvec3 dimensions, size_t formatSizeInBytes,
-                                                    glm::u8 percentageOfAvailableMemory) {
-    uvec3 dim = dimensions;
-
-    // adapt brick size according to available memory
-    while (
-        !canAllocate(getMemorySizeInBytes(dim, formatSizeInBytes), percentageOfAvailableMemory)) {
-        int maxDim = (dim.x > dim.y ? (dim.x > dim.z ? 0 : 2) : (dim.y > dim.z ? 1 : 2));
-        if (dim[maxDim] % 2 != 0) dim[maxDim]++;  // Make the dim we are dividing even
-        dim[maxDim] /= 2;
-    }
-
-    // adapt brick size according to maximum texture dimension
-    unsigned int maxGPUTextureDim = static_cast<unsigned int>(getMaxTexSize());
-
-    while (dim.x > maxGPUTextureDim || dim.y > maxGPUTextureDim || dim.z > maxGPUTextureDim) {
-        int maxDim = (dim.x > dim.y ? (dim.x > dim.z ? 0 : 2) : (dim.y > dim.z ? 1 : 2));
-        if (dim[maxDim] % 2 != 0) dim[maxDim]++;  // Make the dim we are dividing even
-        dim[maxDim] /= 2;
-    }
-
-    return dim;
-}
-
 int OpenGLCapabilities::getOpenGLVersion() { return glVersion_; }
 
-bool OpenGLCapabilities::hasSupportedOpenGLVersion() { return (glVersion_ >= 320); }
+bool OpenGLCapabilities::hasSupportedOpenGLVersion() { return (glVersion_ >= 330); }
 
 bool OpenGLCapabilities::hasOpenGLVersion() { return (glVersion_ > 0); }
 
@@ -262,50 +234,55 @@ void OpenGLCapabilities::initializeGLEW() {
     }
 }
 
+OpenGLCapabilities::GLSLShaderVersion OpenGLCapabilities::getCurrentShaderVersion() {
+    if (supportedShaderVersions_.size() > currentGlobalGLSLVersionIdx_) {
+        return supportedShaderVersions_[currentGlobalGLSLVersionIdx_];
+    } else {
+        return GLSLShaderVersion(0);
+    }
+}
+
+size_t OpenGLCapabilities::getNumberOfShaderVersions() const {
+    return supportedShaderVersions_.size();
+}
+OpenGLCapabilities::GLSLShaderVersion OpenGLCapabilities::getShaderVersion(size_t ind) const {
+    return supportedShaderVersions_[ind];
+}
+
+size_t OpenGLCapabilities::getCurrentShaderIndex() const {
+    return currentGlobalGLSLVersionIdx_;
+}
+
 bool OpenGLCapabilities::isExtensionSupported(const char* name) {
     return (glewIsExtensionSupported(name) != '0');
 }
 
-bool OpenGLCapabilities::isSupported(const char* name) { return (glewIsSupported(name) != '0'); }
-
-bool OpenGLCapabilities::isTexturesSupported() { return texSupported_; }
-
-bool OpenGLCapabilities::isTextureArraysSupported() { return texArraySupported_; }
-
-bool OpenGLCapabilities::is3DTexturesSupported() { return tex3DSupported_; }
-
-bool OpenGLCapabilities::isFboSupported() { return fboSupported_; }
-
-bool OpenGLCapabilities::isShadersSupported() { return shadersAreSupported_; }
-
-bool OpenGLCapabilities::isShadersSupportedARB() { return shadersAreSupportedARB_; }
-
-bool OpenGLCapabilities::isGeometryShadersSupported() { return geometryShsadersAreSupported_; }
-
-OpenGLCapabilities::GLSLShaderVersion OpenGLCapabilities::getCurrentShaderVersion() {
-    if (supportedShaderVersions_.size() > currentGlobalGLSLVersionIdx_)
-        return supportedShaderVersions_[currentGlobalGLSLVersionIdx_];
-    else
-        return 0;
+bool OpenGLCapabilities::isSupported(const char* name) {
+    return (glewIsSupported(name) != '0');
+}
+bool OpenGLCapabilities::isTexturesSupported() const { return texSupported_; }
+bool OpenGLCapabilities::isTextureArraysSupported() const { return texArraySupported_; }
+bool OpenGLCapabilities::is3DTexturesSupported() const { return tex3DSupported_; }
+bool OpenGLCapabilities::isFboSupported() const { return fboSupported_; }
+bool OpenGLCapabilities::isShadersSupported() const { return shadersAreSupported_; }
+bool OpenGLCapabilities::isShadersSupportedARB() const { return shadersAreSupportedARB_; }
+bool OpenGLCapabilities::isGeometryShadersSupported() const {
+    return geometryShsadersAreSupported_;
 }
 
-std::string OpenGLCapabilities::getCurrentGlobalGLSLHeader() {
-    if (currentGlobalGLSLHeader_ == "") rebuildGLSLHeader();
+int OpenGLCapabilities::getMaxProgramLoopCount() const { return maxProgramLoopCount_; }
+int OpenGLCapabilities::getNumTexUnits() const { return numTexUnits_; }
+int OpenGLCapabilities::getMaxTexSize() const { return maxTexSize_; }
+int OpenGLCapabilities::getMax3DTexSize() const { return max3DTexSize_; }
+int OpenGLCapabilities::getMaxArrayTexSize() const { return maxArrayTexSize_; }
+int OpenGLCapabilities::getMaxArrayVertexAttribs() const { return maxArrayVertexAttribs_; }
+int OpenGLCapabilities::getMaxColorAttachments() const { return maxColorAttachments_; }
 
-    return currentGlobalGLSLHeader_;
-}
-
-std::string OpenGLCapabilities::getCurrentGlobalGLSLVertexDefines() {
-    if (currentGlobalGLSLVertexDefines_ == "") rebuildGLSLVertexDefines();
-
-    return currentGlobalGLSLVertexDefines_;
-}
-
-std::string OpenGLCapabilities::getCurrentGlobalGLSLFragmentDefines() {
-    if (currentGlobalGLSLFragmentDefines_ == "") rebuildGLSLFragmentDefines();
-
-    return currentGlobalGLSLFragmentDefines_;
-}
+const std::string& OpenGLCapabilities::getRenderString() const { return glRenderStr_; }
+const std::string& OpenGLCapabilities::getVendorString() const { return glVendorStr_; }
+const std::string& OpenGLCapabilities::getGLVersionString() const { return glVersionStr_; }
+const std::string& OpenGLCapabilities::getGLSLVersionString() const { return glslVersionStr_; }
+OpenGLCapabilities::GlVendor OpenGLCapabilities::getVendor() const { return glVendor_; }
 
 glm::u64 OpenGLCapabilities::getCurrentAvailableTextureMem() {
     glm::u64 currentAvailableTexMeminBytes = 0;
@@ -353,7 +330,8 @@ glm::u64 OpenGLCapabilities::getTotalAvailableTextureMem() {
             wglGetGPUIDsAMD(n, ids);
             wglGetGPUInfoAMD(ids[0], WGL_GPU_RAM_AMD, GL_UNSIGNED_INT, sizeof(size_t),
                              &total_mem_mb);
-            totalAvailableTexMemInBytes = util::megabytes_to_bytes(static_cast<glm::u64>(total_mem_mb));
+            totalAvailableTexMemInBytes =
+                util::megabytes_to_bytes(static_cast<glm::u64>(total_mem_mb));
 #elif defined(GLX_AMD_gpu_association)
             UINT n = glXGetGPUIDsAMD(0, 0);
             UINT* ids = new UINT[n];
@@ -361,7 +339,8 @@ glm::u64 OpenGLCapabilities::getTotalAvailableTextureMem() {
             glXGetGPUIDsAMD(n, ids);
             glXGetGPUInfoAMD(ids[0], GLX_GPU_RAM_AMD, GL_UNSIGNED_INT, sizeof(size_t),
                              &total_mem_mb);
-            totalAvailableTexMemInBytes = util::megabytes_to_bytes(static_cast<glm::u64>(total_mem_mb));
+            totalAvailableTexMemInBytes =
+                util::megabytes_to_bytes(static_cast<glm::u64>(total_mem_mb));
 #endif
         }
     } catch (const Exception& e) {
@@ -370,20 +349,6 @@ glm::u64 OpenGLCapabilities::getTotalAvailableTextureMem() {
 
     return totalAvailableTexMemInBytes;
 }
-
-int OpenGLCapabilities::getMaxProgramLoopCount() { return maxProgramLoopCount_; }
-
-int OpenGLCapabilities::getNumTexUnits() { return numTexUnits_; }
-
-int OpenGLCapabilities::getMaxTexSize() { return maxTexSize_; }
-
-int OpenGLCapabilities::getMax3DTexSize() { return max3DTexSize_; }
-
-int OpenGLCapabilities::getMaxArrayTexSize() { return maxArrayTexSize_; }
-
-int OpenGLCapabilities::getMaxArrayVertexAttribs() { return maxArrayVertexAttribs_; }
-
-int OpenGLCapabilities::getMaxColorAttachments() { return maxColorAttachments_; }
 
 std::string OpenGLCapabilities::getPreferredProfile() { return preferredProfile_; }
 
@@ -421,10 +386,10 @@ void OpenGLCapabilities::retrieveStaticInfo() {
     if (glVendorStr_.find("NVIDIA") != std::string::npos) {
         glVendor_ = GlVendor::Nvidia;
     } else if (glVendorStr_.find("AMD") != std::string::npos ||
-             glVendorStr_.find("ATI") != std::string::npos) {
+               glVendorStr_.find("ATI") != std::string::npos) {
         glVendor_ = GlVendor::Amd;
     } else if (glVendorStr_.find("INTEL") != std::string::npos ||
-             glVendorStr_.find("Intel") != std::string::npos) {
+               glVendorStr_.find("Intel") != std::string::npos) {
         glVendor_ = GlVendor::Intel;
     } else {
         glVendor_ = GlVendor::Unknown;
@@ -533,7 +498,7 @@ void OpenGLCapabilities::retrieveStaticInfo() {
         glGetProgramivARB(GL_FRAGMENT_PROGRAM_ARB, GL_MAX_PROGRAM_LOOP_COUNT_NV, &i);
 
         if (i > 0) {
-            // Restrict cycles to realistic samplingRate*maximumDimension, 
+            // Restrict cycles to realistic samplingRate*maximumDimension,
             // 20*(10 000) slices = 200000
             maxProgramLoopCount_ = std::min<int>(static_cast<int>(i), 200000);
         }
@@ -604,63 +569,6 @@ void OpenGLCapabilities::retrieveStaticInfo() {
 
 void OpenGLCapabilities::retrieveDynamicInfo() {}
 
-void OpenGLCapabilities::rebuildGLSLHeader() {
-    auto current = supportedShaderVersions_[currentGlobalGLSLVersionIdx_];
-    std::stringstream ss;
-
-    ss << "#version " + current.getVersionAndProfileAsString() + "\n";
-
-    if (current.getVersion() == 140 && preferredProfile_ == "compatibility")
-        ss << "#extension GL_ARB_compatibility : enable\n";
-
-    if (current.hasProfile()) {
-        ss << "#define GLSL_PROFILE_" + toUpper(current.getProfile()) + "\n";
-    }
-
-    if (current.getVersion() < 150) {
-        ss << "#define texture(t, s) texture2D(t, s)\n";
-        ss << "#define texture(t, s) texture3D(t, s)\n";
-    }
-
-    int lastVersion = -1;
-
-    for (size_t i = currentGlobalGLSLVersionIdx_; i < supportedShaderVersions_.size(); i++) {
-        if (lastVersion != supportedShaderVersions_[i].getVersion()) {
-            ss << "#define GLSL_VERSION_" + supportedShaderVersions_[i].getVersionAsString() + "\n";
-            lastVersion = supportedShaderVersions_[i].getVersion();
-        }
-    }
-
-    if (getMaxProgramLoopCount() > 0) {
-        ss << "#define MAX_PROGRAM_LOOP_COUNT " + toString(getMaxProgramLoopCount()) + "\n";
-    }
-
-    currentGlobalGLSLHeader_ = ss.str();
-}
-
-void OpenGLCapabilities::rebuildGLSLVertexDefines() {
-    currentGlobalGLSLVertexDefines_ = "";
-
-    // layout locations for GLSL >= 3.30 are handled now via glBindAttribLocation
-    currentGlobalGLSLVertexDefines_ += "in vec4 in_Vertex;\n";
-    currentGlobalGLSLVertexDefines_ += "in vec3 in_Normal;\n";
-    currentGlobalGLSLVertexDefines_ += "in vec4 in_Color;\n";
-    currentGlobalGLSLVertexDefines_ += "in vec3 in_TexCoord;\n";
-}
-
-void OpenGLCapabilities::rebuildGLSLFragmentDefines() {
-    currentGlobalGLSLFragmentDefines_ = "";
-    // layout locations for GLSL >= 3.30 are now handled via glFragDataLocation
-    if (supportedShaderVersions_[currentGlobalGLSLVersionIdx_].getVersion() >= 130) {
-        currentGlobalGLSLFragmentDefines_ += "layout(location = 0) out vec4 FragData0;\n";
-        currentGlobalGLSLFragmentDefines_ += "layout(location = 1) out vec4 PickingData;\n";
-    } else {
-        currentGlobalGLSLFragmentDefines_ += "#define FragData0 gl_FragColor\n";
-        currentGlobalGLSLFragmentDefines_ +=
-            "#define PickingData gl_FragData[" + toString(getMaxColorAttachments() - 1) + "]\n";
-    }
-}
-
 void OpenGLCapabilities::addShaderVersion(GLSLShaderVersion version) {
     supportedShaderVersions_.push_back(version);
 }
@@ -674,10 +582,11 @@ void OpenGLCapabilities::parseAndAddShaderVersion(std::string versionStr, int co
     // Assumes <version><space><profile> or <version>, example 420 core or 140
     if (!versionStr.empty()) {
         // Remove all non-alphanumeric characters, but keep spaces
-        versionStr.erase(std::remove_if(versionStr.begin(), versionStr.end(), [](char c) {
-                             return !(std::isspace(c) || std::isalnum(c));
-                         }), versionStr.end());
-        
+        versionStr.erase(
+            std::remove_if(versionStr.begin(), versionStr.end(),
+                           [](char c) { return !(std::isspace(c) || std::isalnum(c)); }),
+            versionStr.end());
+
         auto versionSplit = splitString(versionStr);
         if (versionSplit.size() > 1 && (versionSplit[1].compare("core") == 0 ||
                                         versionSplit[1].compare("compatibility") == 0)) {
@@ -708,10 +617,4 @@ int OpenGLCapabilities::parseAndRetrieveVersion(std::string versionStr) {
     return 0;
 }
 
-std::string OpenGLCapabilities::getRenderString() const { return glRenderStr_; }
-std::string OpenGLCapabilities::getVendorString() const { return glVendorStr_; }
-std::string OpenGLCapabilities::getGLVersionString() const { return glVersionStr_; }
-std::string OpenGLCapabilities::getGLSLVersionString() const { return glslVersionStr_; }
-OpenGLCapabilities::GlVendor OpenGLCapabilities::getVendor() const { return glVendor_; }
-
-}  // namespace
+}  // namespace inviwo

--- a/modules/opengl/src/shader/shader.cpp
+++ b/modules/opengl/src/shader/shader.cpp
@@ -105,10 +105,10 @@ Shader::Shader(const std::vector<std::pair<ShaderType, std::string>> &items, Bui
     : warningLevel_{UniformWarning::Ignore} {
 
     for (auto &item : items) {
-        shaderObjects_.insert(
-            {item.first,
-             ShaderAttachment(this, std::make_unique<ShaderObject>(
-                                        item.first, utilgl::findShaderResource(item.second)))});
+        shaderObjects_.emplace(
+            item.first,
+            ShaderAttachment(this, std::make_unique<ShaderObject>(
+                                       item.first, utilgl::findShaderResource(item.second))));
     }
 
     verify();
@@ -122,8 +122,8 @@ Shader::Shader(
     : warningLevel_{UniformWarning::Ignore} {
 
     for (auto &item : items) {
-        shaderObjects_.insert({item.first, ShaderAttachment(this, std::make_unique<ShaderObject>(
-                                                                      item.first, item.second))});
+        shaderObjects_.emplace(item.first, ShaderAttachment(this, std::make_unique<ShaderObject>(
+                                                                      item.first, item.second)));
     }
     verify();
     if (buildShader == Build::Yes) build();
@@ -135,7 +135,7 @@ Shader::Shader(std::vector<std::unique_ptr<ShaderObject>> &shaderObjects, bool b
 
     for (auto &obj : shaderObjects) {
         auto type = obj->getShaderType();
-        shaderObjects_.insert({type, ShaderAttachment(this, std::move(obj))});
+        shaderObjects_.emplace(type, ShaderAttachment(this, std::move(obj)));
     }
 
     verify();
@@ -178,8 +178,8 @@ Shader::Shader(const char *vertexFilename, const char *geometryFilename,
 
 Shader::Shader(const Shader &rhs) : program_{rhs.program_}, warningLevel_{rhs.warningLevel_} {
     for (auto &elem : rhs.shaderObjects_) {
-        shaderObjects_[elem.first] =
-            ShaderAttachment(this, util::make_unique<ShaderObject>(elem.second.obj()));
+        shaderObjects_.emplace(
+            elem.first, ShaderAttachment(this, util::make_unique<ShaderObject>(elem.second.obj())));
     }
 
     if (rhs.isReady()) build();
@@ -209,8 +209,9 @@ Shader &Shader::operator=(const Shader &that) {
 
         shaderObjects_.clear();
         for (auto &elem : that.shaderObjects_) {
-            shaderObjects_[elem.first] =
-                ShaderAttachment(this, util::make_unique<ShaderObject>(elem.second.obj()));
+            shaderObjects_.emplace(
+                elem.first,
+                ShaderAttachment(this, util::make_unique<ShaderObject>(elem.second.obj())));
         }
         warningLevel_ = that.warningLevel_;
 
@@ -271,7 +272,7 @@ void Shader::link() {
 
 void Shader::linkShader(bool notifyRebuild) {
     uniformLookup_.clear();  // clear uniform location cache.
-    ShaderManager::getPtr()->bindCommonAttributes(program_.id);
+    bindAttributes();
 
     if (!util::all_of(shaderObjects_,
                       [](const auto &elem) { return elem.second.obj().isReady(); })) {
@@ -299,6 +300,17 @@ void Shader::linkShader(bool notifyRebuild) {
     LGL_ERROR;
     ready_ = true;
     if (notifyRebuild) onReloadCallback_.invokeAll();
+}
+
+void Shader::bindAttributes() {
+    for (const auto &obj : getShaderObjects()) {
+        for (const auto &item : obj.getInDeclarations()) {
+            glBindAttribLocation(program_.id, item.location, item.name.c_str());
+        }
+        for (const auto &item : obj.getOutDeclarations()) {
+            glBindFragDataLocation(program_.id, item.location, item.name.c_str());
+        }
+    }
 }
 
 void Shader::rebuildShader(ShaderObject *obj) {
@@ -430,7 +442,7 @@ ShaderObject *Shader::operator[](ShaderType type) const { return getShaderObject
 ShaderObject *Shader::getShaderObject(ShaderType type) const {
     auto it = shaderObjects_.find(type);
     if (it != shaderObjects_.end()) {
-       return &(it->second.obj()); 
+        return &(it->second.obj());
     } else {
         return nullptr;
     }

--- a/modules/opengl/src/shader/shadermanager.cpp
+++ b/modules/opengl/src/shader/shadermanager.cpp
@@ -85,37 +85,9 @@ bool ShaderManager::isRegistered(Shader* shader) const {
     return std::find(shaders_.begin(), shaders_.end(), shader) != shaders_.end();
 }
 
-std::string ShaderManager::getGlobalGLSLHeader() {
-    OpenGLCapabilities* glCaps = getOpenGLCapabilitiesObject();
-    return glCaps->getCurrentGlobalGLSLHeader();
-}
-
-std::string ShaderManager::getGlobalGLSLVertexDefines() {
-    OpenGLCapabilities* glCaps = getOpenGLCapabilitiesObject();
-    return glCaps->getCurrentGlobalGLSLVertexDefines();
-}
-
-std::string ShaderManager::getGlobalGLSLFragmentDefines() {
-    OpenGLCapabilities* glCaps = getOpenGLCapabilitiesObject();
-    return glCaps->getCurrentGlobalGLSLFragmentDefines();
-}
-
 int ShaderManager::getGlobalGLSLVersion() {
-    OpenGLCapabilities* glCaps = getOpenGLCapabilitiesObject();
+    OpenGLCapabilities* glCaps = getOpenGLCapabilities();
     return glCaps->getCurrentShaderVersion().getVersion();
-}
-
-void ShaderManager::bindCommonAttributes(unsigned int programID) {
-    int glslVersion = this->getGlobalGLSLVersion();
-    glBindAttribLocation(programID, 0, "in_Vertex");
-    glBindAttribLocation(programID, 1, "in_Normal");
-    glBindAttribLocation(programID, 2, "in_Color");
-    glBindAttribLocation(programID, 3, "in_TexCoord");
-
-    if (glslVersion >= 130) {
-        glBindFragDataLocation(programID, 0, "FragData0");
-        glBindFragDataLocation(programID, 1, "PickingData");
-    }
 }
 
 const std::vector<std::string>& ShaderManager::getShaderSearchPaths() { return shaderSearchPaths_; }
@@ -181,7 +153,7 @@ std::shared_ptr<ShaderResource> ShaderManager::getShaderResource(std::string key
     return nullptr;
 }
 
-OpenGLCapabilities* ShaderManager::getOpenGLCapabilitiesObject() {
+OpenGLCapabilities* ShaderManager::getOpenGLCapabilities() {
     if (!openGLInfoRef_) {
         if (auto openGLModule = InviwoApplication::getPtr()->getModuleByType<OpenGLModule>())
             openGLInfoRef_ = getTypeFromVector<OpenGLCapabilities>(openGLModule->getCapabilities());

--- a/modules/opengl/src/shader/shaderobject.cpp
+++ b/modules/opengl/src/shader/shaderobject.cpp
@@ -38,8 +38,21 @@
 #include <modules/opengl/openglexception.h>
 #include <modules/opengl/shader/shadermanager.h>
 #include <modules/opengl/shader/shaderutils.h>
+#include <modules/opengl/openglcapabilities.h>
+
+#include <fmt/format.h>
 
 namespace inviwo {
+
+std::string ShaderObject::InDeclaration::toString() const {
+    return fmt::format(decl, fmt::arg("type", type), fmt::arg("name", name),
+                       fmt::arg("location", location));
+}
+
+std::string ShaderObject::OutDeclaration::toString() const {
+    return fmt::format(decl, fmt::arg("type", type), fmt::arg("name", name),
+                       fmt::arg("location", location));
+}
 
 ShaderObject::ShaderObject(ShaderType shaderType, std::shared_ptr<const ShaderResource> resource)
     : shaderType_{shaderType}, id_{glCreateShader(shaderType)}, resource_{resource} {
@@ -52,6 +65,12 @@ ShaderObject::ShaderObject(ShaderType shaderType, std::shared_ptr<const ShaderRe
     std::string fileExtension = filesystem::getFileExtension(resource_->key());
     if (fileExtension != shaderType_.extension()) {
         LogWarn("File extension does not match shader type: " << resource_->key());
+    }
+
+    if (shaderType_ == ShaderType::Fragment) {
+        addStandardFragmentOutDeclarations();
+    } else if (shaderType_ == ShaderType::Vertex) {
+        addStandardVertexInDeclarations();
     }
 }
 
@@ -68,32 +87,31 @@ ShaderObject::ShaderObject(std::string fileName)
 ShaderObject::ShaderObject(GLenum shaderType, std::string fileName)
     : ShaderObject(ShaderType(shaderType), loadResource(fileName)) {}
 
-
 ShaderObject::ShaderObject(const ShaderObject& rhs)
     : shaderType_(rhs.shaderType_)
     , id_(glCreateShader(rhs.shaderType_))
     , resource_(rhs.resource_)
+    , inDeclarations_(rhs.inDeclarations_)
     , outDeclarations_(rhs.outDeclarations_)
     , shaderDefines_(rhs.shaderDefines_)
     , shaderExtensions_(rhs.shaderExtensions_)
     , sourceProcessed_{}
-    , includeResources_{} 
-    , lineNumberResolver_{} 
+    , includeResources_{}
+    , lnr_{}
     , callbacks_{}
-    , resourceCallbacks_{} {
-}
-
+    , resourceCallbacks_{} {}
 
 ShaderObject::ShaderObject(ShaderObject&& rhs) noexcept
     : shaderType_(rhs.shaderType_)
     , id_(rhs.id_)
     , resource_(std::move(rhs.resource_))
+    , inDeclarations_(std::move(rhs.inDeclarations_))
     , outDeclarations_(std::move(rhs.outDeclarations_))
     , shaderDefines_(std::move(rhs.shaderDefines_))
     , shaderExtensions_(std::move(rhs.shaderExtensions_))
     , sourceProcessed_{}
-    , includeResources_{} 
-    , lineNumberResolver_{} 
+    , includeResources_{}
+    , lnr_{}
     , callbacks_{}
     , resourceCallbacks_{} {
 
@@ -106,12 +124,13 @@ ShaderObject& ShaderObject::operator=(const ShaderObject& that) {
     std::swap(shaderType_, copy.shaderType_);
     std::swap(id_, copy.id_);
     std::swap(resource_, copy.resource_);
+    std::swap(inDeclarations_, copy.inDeclarations_);
     std::swap(outDeclarations_, copy.outDeclarations_);
     std::swap(shaderDefines_, copy.shaderDefines_);
     std::swap(shaderExtensions_, copy.shaderExtensions_);
     std::swap(sourceProcessed_, copy.sourceProcessed_);
     std::swap(includeResources_, copy.includeResources_);
-    std::swap(lineNumberResolver_, copy.lineNumberResolver_);
+    std::swap(lnr_, copy.lnr_);
     std::swap(callbacks_, copy.callbacks_);
     std::swap(resourceCallbacks_, copy.resourceCallbacks_);
     return *this;
@@ -122,12 +141,13 @@ ShaderObject& ShaderObject::operator=(ShaderObject&& that) noexcept {
     std::swap(shaderType_, copy.shaderType_);
     std::swap(id_, copy.id_);
     std::swap(resource_, copy.resource_);
+    std::swap(inDeclarations_, copy.inDeclarations_);
     std::swap(outDeclarations_, copy.outDeclarations_);
     std::swap(shaderDefines_, copy.shaderDefines_);
     std::swap(shaderExtensions_, copy.shaderExtensions_);
     std::swap(sourceProcessed_, copy.sourceProcessed_);
     std::swap(includeResources_, copy.includeResources_);
-    std::swap(lineNumberResolver_, copy.lineNumberResolver_);
+    std::swap(lnr_, copy.lnr_);
     std::swap(callbacks_, copy.callbacks_);
     std::swap(resourceCallbacks_, copy.resourceCallbacks_);
     return *this;
@@ -159,67 +179,73 @@ void ShaderObject::build() {
 
 void ShaderObject::preprocess() {
     resourceCallbacks_.clear();
-    lineNumberResolver_.clear();
+    lnr_.clear();
     auto holdOntoResources = includeResources_;  // Don't release until we have processed again.
     includeResources_.clear();
 
     std::ostringstream source;
     addDefines(source);
-    addOutDeclarations(source);
     addIncludes(source, resource_);
     sourceProcessed_ = source.str();
 }
 
 void ShaderObject::addDefines(std::ostringstream& source) {
-    {
-        std::string globalGLSLHeader = ShaderManager::getPtr()->getGlobalGLSLHeader();
-        std::string curLine;
-        std::istringstream globalGLSLHeaderStream(globalGLSLHeader);
-        while (std::getline(globalGLSLHeaderStream, curLine)) {
-            lineNumberResolver_.emplace_back("GlobalGLSLSHeader", 0);
-        }
-        source << globalGLSLHeader;
+    auto capa = ShaderManager::getPtr()->getOpenGLCapabilities();
+    const auto current = capa->getCurrentShaderVersion();
+
+    source << "#version " << current.getVersionAndProfileAsString() << "\n";
+    lnr_.addLine("Version", 0);
+
+    for (const auto& se : shaderExtensions_) {
+        source << "#extension " << se.first << " : " << (se.second ? "enable" : "disable") << "\n";
+        lnr_.addLine("Extensions", 0);
     }
 
-    {
-        for (const auto& se : shaderExtensions_) {
-            source << "#extension " << se.first << " : " << (se.second ? "enable" : "disable")
-                   << "\n";
-            lineNumberResolver_.emplace_back("Extension", 0);
-        }
+    if (current.hasProfile()) {
+        source << "#define GLSL_PROFILE_" + toUpper(current.getProfile()) + "\n";
+        lnr_.addLine("Header", 0);
     }
-    {
-        for (const auto& sd : shaderDefines_) {
-            source << "#define " << sd.first << " " << sd.second << "\n";
-            lineNumberResolver_.emplace_back("Defines", 0);
+
+    int lastVersion = -1;
+    for (size_t i = capa->getCurrentShaderIndex(); i < capa->getNumberOfShaderVersions(); i++) {
+        const auto version = capa->getShaderVersion(i);
+        if (lastVersion != version.getVersion()) {
+            source << "#define GLSL_VERSION_" << version.getVersionAsString() << "\n";
+            lnr_.addLine("Header", 0);
+            lastVersion = version.getVersion();
         }
     }
 
-    {
-        std::string globalDefines;
-        if (shaderType_ == ShaderType::Vertex) {
-            globalDefines += ShaderManager::getPtr()->getGlobalGLSLVertexDefines();
-        } else if (shaderType_ == ShaderType::Fragment) {
-            globalDefines += ShaderManager::getPtr()->getGlobalGLSLFragmentDefines();
-        }
+    if (capa->getMaxProgramLoopCount() > 0) {
+        source << "#define MAX_PROGRAM_LOOP_COUNT " << capa->getMaxProgramLoopCount() << "\n";
+        lnr_.addLine("Header", 0);
+    }
 
-        std::string curLine;
-        std::istringstream globalGLSLDefinesStream(globalDefines);
-        while (std::getline(globalGLSLDefinesStream, curLine)) {
-            lineNumberResolver_.emplace_back("GlobalGLSLSDefines", 0);
-        }
-        source << globalDefines;
+    for (const auto& sd : shaderDefines_) {
+        source << "#define " << sd.first << " " << sd.second << "\n";
+        lnr_.addLine("Defines", 0);
+    }
+
+    for (auto decl : outDeclarations_) {
+        source << decl.toString() << "\n";
+        lnr_.addLine("Out Declaration", 0);
+    }
+    for (auto decl : inDeclarations_) {
+        source << decl.toString() << "\n";
+        lnr_.addLine("In Declaration", 0);
     }
 }
 
-void ShaderObject::addOutDeclarations(std::ostringstream& source) {
-    for (auto curDeclaration : outDeclarations_) {
-        if (curDeclaration.second > -1) {
-            source << "layout(location = " << curDeclaration.second << ") ";
-        }
-        source << "out vec4 " << curDeclaration.first << ";\n";
-        lineNumberResolver_.emplace_back("Out Declaration", 0);
-    }
+void ShaderObject::addStandardFragmentOutDeclarations() {
+    addOutDeclaration("FragData0", 0);
+    addOutDeclaration("PickingData", 1);
+}
+
+void ShaderObject::addStandardVertexInDeclarations() {
+    addInDeclaration(InDeclaration{"in_Vertex", 0});
+    addInDeclaration(InDeclaration{"in_Normal", 1, "vec3"});
+    addInDeclaration(InDeclaration{"in_Color", 2});
+    addInDeclaration(InDeclaration{"in_TexCoord", 3, "vec3"});
 }
 
 void ShaderObject::addIncludes(std::ostringstream& source,
@@ -313,7 +339,7 @@ void ShaderObject::addIncludes(std::ostringstream& source,
 
         if (!hasAddedInclude) {  // include the whole line
             source << curLine << "\n";
-            lineNumberResolver_.emplace_back(resource->key(), localLineNumber);
+            lnr_.addLine(resource->key(), localLineNumber);
         }
         localLineNumber++;
     }
@@ -335,17 +361,14 @@ void ShaderObject::compile() {
     glCompileShader(id_);
     if (!isReady()) {
         throw OpenGLException(
-            resource_->key() + " " +
-                utilgl::reformatInfoLog(lineNumberResolver_, utilgl::getShaderInfoLog(id_)),
-            IvwContext);
+            resource_->key() + " " + lnr_.resolveLog(utilgl::getShaderInfoLog(id_)), IVW_CONTEXT);
     }
 
 #ifdef IVW_DEBUG
-    auto log = utilgl::getShaderInfoLog(id_);
+    const auto log = utilgl::getShaderInfoLog(id_);
     if (!log.empty()) {
-        util::log(IvwContext,
-                  resource_->key() + " " + utilgl::reformatInfoLog(lineNumberResolver_, log),
-                  LogLevel::Info, LogAudience::User);
+        util::log(IVW_CONTEXT, resource_->key() + " " + lnr_.resolveLog(log), LogLevel::Info,
+                  LogAudience::User);
     }
 #endif
 }
@@ -374,50 +397,110 @@ bool ShaderObject::hasShaderExtension(const std::string& extName) const {
 
 void ShaderObject::clearShaderExtensions() { shaderExtensions_.clear(); }
 
-void ShaderObject::addOutDeclaration(std::string name, int location) {
+void ShaderObject::addOutDeclaration(std::string name, int location, const std::string& type) {
+    addOutDeclaration(OutDeclaration{name, location, type});
+}
+
+void ShaderObject::addOutDeclaration(const OutDeclaration& decl) {
     auto it = util::find_if(outDeclarations_,
-                            [&](std::pair<std::string, int>& elem) { return elem.first == name; });
+                            [&](const OutDeclaration& elem) { return elem.name == decl.name; });
     if (it != outDeclarations_.end()) {
-        it->second = location;
+        *it = decl;
     } else {
-        outDeclarations_.push_back({name, location});
+        outDeclarations_.push_back(decl);
     }
+}
+
+auto ShaderObject::getOutDeclarations() const -> const std::vector<OutDeclaration>& {
+    return outDeclarations_;
 }
 
 void ShaderObject::clearOutDeclarations() { outDeclarations_.clear(); }
 
-std::pair<std::string, unsigned int> ShaderObject::resolveLine(size_t line) const {
-    if (line < lineNumberResolver_.size())
-        return lineNumberResolver_[line];
-    else
+void ShaderObject::ShaderObject::addInDeclaration(std::string name, int location,
+                                                  const std::string& type) {
+    addInDeclaration(InDeclaration{name, location, type});
+}
+void ShaderObject::addInDeclaration(const InDeclaration& decl) {
+
+    auto it = util::find_if(inDeclarations_,
+                            [&](const InDeclaration& elem) { return elem.name == decl.name; });
+    if (it != inDeclarations_.end()) {
+        *it = decl;
+    } else {
+        inDeclarations_.push_back(decl);
+    }
+}
+void ShaderObject::clearInDeclarations() { inDeclarations_.clear(); }
+auto ShaderObject::getInDeclarations() const -> const std::vector<InDeclaration>& {
+    return inDeclarations_;
+}
+
+std::pair<std::string, size_t> ShaderObject::LineNumberResolver::resolveLine(size_t line) const {
+    if (line < lines_.size()) {
+        return lines_[line];
+    } else {
         return {"", 0};
+    }
+}
+
+void ShaderObject::LineNumberResolver::addLine(const std::string& file, size_t line) {
+    lines_.emplace_back(file, line);
+}
+
+void ShaderObject::LineNumberResolver::clear() { lines_.clear(); }
+
+std::string ShaderObject::LineNumberResolver::resolveLog(const std::string& compileLog) const {
+    std::ostringstream result;
+    std::istringstream origShaderInfoLog(compileLog);
+
+    std::string curLine;
+    while (std::getline(origShaderInfoLog, curLine)) {
+        if (!curLine.empty()) {
+            const int origLineNumber = utilgl::getLogLineNumber(curLine);
+            if (origLineNumber > 0) {
+                const auto res = resolveLine(origLineNumber);
+                result << "\n"
+                       << res.first << " (" << res.second
+                       << "): " << curLine.substr(curLine.find(":") + 1);
+            } else {
+                result << "\n" << curLine;
+            }
+        }
+    }
+
+    return std::move(result).str();
+}
+
+std::pair<std::string, size_t> ShaderObject::resolveLine(size_t line) const {
+    return lnr_.resolveLine(line);
 }
 
 std::string ShaderObject::print(bool showSource, bool preprocess) const {
     if (preprocess) {
         if (showSource) {
             std::string::size_type width = 0;
-            for (auto l : lineNumberResolver_) {
+            for (auto l : lnr_) {
                 std::string file = splitString(l.first, '/').back();
                 width = std::max(width, file.length());
             }
 
             size_t i = 0;
-            std::string line;
             std::stringstream out;
             std::istringstream in(sourceProcessed_);
+
+            std::string line;
             while (std::getline(in, line)) {
-                std::string file = i < lineNumberResolver_.size()
-                                       ? splitString(lineNumberResolver_[i].first, '/').back()
-                                       : "";
-                unsigned int lineNumber =
-                    i < lineNumberResolver_.size() ? lineNumberResolver_[i].second : 0;
+                const auto res = lnr_.resolveLine(i);
+                const std::string file =
+                    res.first.empty() ? "" : splitString(res.first, '/').back();
+                const size_t lineNumber = res.second;
 
                 out << std::left << std::setw(width + 1u) << file << std::right << std::setw(4)
                     << lineNumber << ": " << std::left << line << "\n";
                 ++i;
             }
-            return out.str();
+            return std::move(out).str();
         } else {
             return sourceProcessed_;
         }

--- a/modules/opengl/src/shader/shaderutils.cpp
+++ b/modules/opengl/src/shader/shaderutils.cpp
@@ -592,32 +592,6 @@ int getLogLineNumber(const std::string& compileLogLine) {
     return result;
 }
 
-std::string reformatInfoLog(
-    const std::vector<std::pair<std::string, unsigned int> >& lineNumberResolver,
-    const std::string compileLog) {
-
-    std::ostringstream result;
-    std::string curLine;
-    std::istringstream origShaderInfoLog(compileLog);
-
-    while (std::getline(origShaderInfoLog, curLine)) {
-        if (!curLine.empty()) {
-            int origLineNumber = getLogLineNumber(curLine);
-            if (origLineNumber > 0) {
-                auto lineNumber = lineNumberResolver[origLineNumber - 1].second;
-                auto fileName = lineNumberResolver[origLineNumber - 1].first;
-                result << "\n"
-                       << fileName << " (" << lineNumber
-                       << "): " << curLine.substr(curLine.find(":") + 1);
-            } else {
-                result << "\n" << curLine;
-            }
-        }
-    }
-
-    return result.str();
-}
-
 std::string getShaderInfoLog(GLuint id) {
     GLint maxLogLength;
     glGetShaderiv(id, GL_INFO_LOG_LENGTH, &maxLogLength);

--- a/src/core/CMakeLists.txt
+++ b/src/core/CMakeLists.txt
@@ -661,6 +661,7 @@ target_link_libraries(inviwo-core PUBLIC
     inviwo::tinydir
     inviwo::tclap
     inviwo::warn
+    fmt::fmt
     $<$<BOOL:${IVW_SIGAR}>:inviwo::sigar>
     $<$<BOOL:${WIN32}>:inviwo::stackwalker>
     $<$<BOOL:${UNIX}>:${CMAKE_DL_LIBS}>  # Required for dlopen

--- a/src/core/util/capabilities.cpp
+++ b/src/core/util/capabilities.cpp
@@ -34,17 +34,4 @@ namespace inviwo {
 Capabilities::Capabilities() = default;
 Capabilities::~Capabilities() = default;
 
-bool Capabilities::canAllocate(glm::u64 /*dataSize*/, glm::u8 /*percentageOfAvailableMemory*/) {
-    return true;
-}
-
-uvec3 Capabilities::calculateOptimalBrickSize(uvec3 dimensions, size_t /*formatSizeInBytes*/,
-                                              glm::u8 /*percentageOfAvailableMemory*/) {
-    return dimensions;
-}
-
-glm::u64 Capabilities::getMemorySizeInBytes(uvec3 dimensions, size_t formatSizeInBytes) {
-    return static_cast<glm::u64>(dimensions.x * dimensions.y * dimensions.z * formatSizeInBytes);
-}
-
 }  // namespace

--- a/src/core/util/systemcapabilities.cpp
+++ b/src/core/util/systemcapabilities.cpp
@@ -56,26 +56,6 @@ SystemCapabilities::~SystemCapabilities() {
 #endif
 }
 
-bool SystemCapabilities::canAllocate(glm::u64 dataSize, glm::u8 percentageOfAvailableMemory) {
-    return getAvailableMemory() * percentageOfAvailableMemory / 100 >= dataSize;
-}
-
-uvec3 SystemCapabilities::calculateOptimalBrickSize(uvec3 dimensions, size_t formatSizeInBytes,
-                                                    glm::u8 percentageOfAvailableMemory) {
-    uvec3 dim = dimensions;
-
-    while (
-        !canAllocate(getMemorySizeInBytes(dim, formatSizeInBytes), percentageOfAvailableMemory)) {
-        int theMaxDim = (dim.x > dim.y ? (dim.x > dim.z ? 0 : 2) : (dim.y > dim.z ? 1 : 2));
-
-        if (dim[theMaxDim] % 2 != 0) dim[theMaxDim]++;  // Make the dim we are dividing even
-
-        dim[theMaxDim] /= 2;
-    }
-
-    return dim;
-}
-
 void SystemCapabilities::retrieveStaticInfo() {
     successOSInfo_ = lookupOSInfo();
     buildInfo_ = util::getBuildInfo();

